File: Python/CircularQueueArray.py
class CircularQueueArray:

  def __init__(self, n):
    self.MAX_SIZE = n
    self.queue = [0]*n
    self.rear = -1
    self.front = 0
    self.currentSize = 0

  def isFull(self):
    return self.currentSize == self.MAX_SIZE

  def isEmpty(self):
    return self.currentSize == 0

  def enqueue(self, data):
    if self.isFull():
      print("Enqueue Process Failed.\nQueue is at Maximum Capacity!")
      return
    print(data, " added to the back of the queue")

    self.rear = (self.rear+1)%self.MAX_SIZE
    self.queue[self.rear] = data
    self.currentSize+=1

  def dequeue(self):
    if self.isEmpty():
      print("Dequeue Process Failed.\nQueue is Empty!")
      return -self.MAX_SIZE
    data = self.queue[self.front]
    print(data, " removed from the front of the queue")

    self.front = (self.front+1)%self.MAX_SIZE
    self.currentSize-=1
    return data

  def frontElement(self):
    if self.isEmpty():
      print("Unable to get data.\nQueue is Empty!")
      return -self.MAX_SIZE
    return self.queue[self.front]

  def rearElement(self):
    if self.isEmpty():
      print("Unable to get data.\nQueue is Empty!")
      return -self.MAX_SIZE
    return self.queue[self.rear]


def main():
  queue = CircularQueueArray(3)
  print("Is the queue empty?\t", queue.isEmpty())
  queue.dequeue()
  queue.enqueue(10)
  queue.enqueue(20)
  queue.enqueue(40)
  print("Element at the front of the queue: ", queue.frontElement())
  print("Element at the rear of the queue: ", queue.rearElement())
  print("Is the queue full?\t", queue.isFull())
  queue.enqueue(80)
  queue.dequeue()
  queue.enqueue(80)
  print("Is the queue full?\t", queue.isFull())
  print("Is the queue empty?\t", queue.isEmpty())


main()

File: Java/CircularQueueArray.java
class CircularQueueArray {
  private int[] queue;
  private int rear, front, currentSize;
  private final int MAX_SIZE;
  CircularQueueArray(int n) {
    MAX_SIZE = n;
    queue = new int[n];
    rear = -1;
    front = currentSize = 0;
  }
  boolean isFull() {
    return (currentSize == MAX_SIZE);
  }
  boolean isEmpty() {
    return (currentSize == 0);
  }
  void enqueue(int data) {
    if (isFull()) {
      System.out.println("Enqueue Process Failed.\nQueue is at Maximum Capacity!");
      return;
    }
    System.out.println(data + " added to the back of the queue");

    rear = (rear + 1) % MAX_SIZE;
    queue[rear] = data;
    currentSize++;
  }
  int dequeue() {
    if (isEmpty()) {
      System.out.println("Dequeue Process Failed.\nQueue is Empty!");
      return -MAX_SIZE;
    }
    int data = queue[front];
    System.out.println(data + " removed from the front of the queue");

    front = (front + 1) % MAX_SIZE;
    currentSize--;
    return data;
  }
  int front() {
    if (isEmpty()) {
      System.out.println("Unable to get data.\nQueue is Empty!");
      return -MAX_SIZE;
    }
    return queue[front];
  }
  int rear() {
    if (isEmpty()) {
      System.out.println("Unable to get data.\nQueue is Empty!");
      return Integer.MIN_VALUE;
    }
    return queue[rear];
  }
  
  public static void main(String[] args) {
    CircularQueueArray queue = new CircularQueueArray(3);
    System.out.println("Is the queue empty?\t" + queue.isEmpty());
    queue.dequeue();
    queue.enqueue(10);
    queue.enqueue(20);
    queue.enqueue(40);
    System.out.println("Element at the front of the queue: " + queue.front());
    System.out.println("Element at the rear of the queue: " + queue.rear());
    System.out.println("Is the queue full?\t" + queue.isFull());
    queue.enqueue(80);
    queue.dequeue();
    queue.enqueue(80);
    System.out.println("Is the queue full?\t" + queue.isFull());
    System.out.println("Is the queue empty?\t" + queue.isEmpty());

  }
}